components/places/android/src/main/java/org/mozilla/places/Bookmarks.kt

```diff
@@ -0,0 +1,407 @@
+ * (which have no parent), and it's 4 children (which have the
+ * actual root as their parent).
+ *
+ * You cannot delete or move any of these items.
+ */
+enum class BookmarkRoot(val id: String) {
+    Root("root________"),
+    Menu("menu________"),
+    Toolbar("toolbar_____"),
+    Unfiled("unfiled_____"),
+    Mobile("mobile______"),
+}
+
+/**
+ * Enumeration of the type of a bookmark item.
+ *
+ * Must match BookmarkType in the Rust code.
+ */
+enum class BookmarkType(val value: Int) {
+    Bookmark(1),
+    Folder(2),
+    Separator(3),
+}
+
+/**
+ * An interface defining the set of fields common to all nodes
+ * in the bookmark tree.
+ */
+interface BookmarkTreeNode {
+    /**
+     * The type of this bookmark.
+     */
+    val type: BookmarkType
+
+    /**
+     * The guid of this record. Bookmark guids are always 12 characters in the url-safe
+     * base64 character set.
+     */
+    val guid: String
+
+    /**
+     * Creation time, in milliseconds since the unix epoch.
+     *
+     * May not be a local timestamp.
+     */
+    val dateAdded: Long
+
+    /**
+     * Last modification time, in milliseconds since the unix epoch.
+     *
+     * May not be a local timestamp.
+     */
+    val lastModified: Long
+
+    /**
+     * The guid of this record's parent. It should only be null for
+     * [BookmarkRoot.Root].
+     */
+    val parentGUID: String?
+
+    /**
+     * The (0-based) position of this record within it's parent.
+     */
+    val position: Int
+}
+
+/**
+ * A bookmark tree node that actually represents a bookmark.
+ *
+ * It's type is always [BookmarkType.Bookmark], and it has a `title `and `url`
+ * in addition to the fields defined by [BookmarkTreeNode].
+ */
+
+open class BookmarkItem(
+        override val guid: String,
+        override val dateAdded: Long,
+        override val lastModified: Long,
+        override val parentGUID: String?,
+        override val position: Int,
+
+        /**
+         * The URL of this bookmark.
+         */
+        val url: String,
+
+        /**
+         * The title of the bookmark, if any was provided.
+         */
+        val title: String?
+) : BookmarkTreeNode {
+    override val type get() = BookmarkType.Bookmark
+}
+
+/**
+ * A bookmark which is a folder.
+ *
+ * It's type is always [BookmarkType.Folder], and it has a `title`,
+ * a list of `childGUIDs`, and possibly a list of `children` in
+ * addition to those defined by [BookmarkTreeNode].
+ */
+open class BookmarkFolder(
+        override val guid: String,
+        override val dateAdded: Long,
+        override val lastModified: Long,
+        override val parentGUID: String?,
+        override val position: Int,
+
+        /**
+         * The title of this bookmark folder, if any was provided.
+         */
+        val title: String?,
+
+        /**
+         * The GUIDs of this folder's list of children.
+         */
+        val childGUIDs: List<String>,
```

Comment:
Why do we need both `childGUIDs` and `children`? Can't `childGUIDs` just be a map (or maybe filter+map) function over `children`?

Reply:
Oh, so it seems like this split is essentially a perf optimization?

Reply:
Yes, we don't want to return the entire subtree if not needed. If you don't care about this for android we can remove the distinction.